### PR TITLE
(BSR) refactor(dms): Update 'Champs' Handling

### PR DESCRIPTION
**Pull request by dcuesta-pass**
# DE PR

## Describe your changes

Please include a summary of the changes:

### Type of change

* [x] Fix (non-breaking change which corrects expected behavior)
* [ ] New fields (non-breaking change)
* [ ] New table (non-breaking change)
* [ ] Concept change (potentially breaking change which modifies fields according to new or evolving business concepts)
* [ ] Table deletion (potentially breaking change which adds functionality/ table)

### Checklist before requesting a review

* [x] I have performed a self-review of my code

* [x] My code passes CI/CD tests
* [x] I updated README.md
* [x] I have updated the dag
* [x] I will create a review on slack and ensure to specify the duration of the review task: short (<10min), medium (<30min), long (>30min)

### Added tests?

* [ ] 👍 yes
* [x] 🙅 no, because they aren't needed
* [ ] 🙋 no, because I need help
* [ ] ⏰ no, but I created a ticket

### PR title format (except for MEP)

There is a linter on the PR title format. Please respect the following format:

<details>
<summary>(ticket) type(topic): comment</summary>

* ticket surrounded by parenthesis, with optionnaly a hyphen followed by one or more digits (e.g., -1234). The first part must be one of the following strings:
  * DA
  * DE
  * AE
  * DS
  * HF
  * BSR
  * PC

* type :
The second part to specify the type of change one of the following :
  * build
  * lint
  * ci
  * docs
  * feat
  * fix
  * perf
  * refactor
  * test
  * core
  * dbt

* topic within parenthesis: 1 word e.g., (dag)

* comment: tell us your life

examples:

* :white_check_mark: (DE-124) refactor(firebase): update source field
* :x: (DE-124) refactor (firebase): update source field **(space between type and topic)**
* :x: (DE-124) airflow(firebase): update source fiedd in DAG **(wrong type)**
* :x: (DE-124) (DE-124) refactor(firebase refacto): update source field **(topic in two words)**
* :white_check_mark: (BSR) docs(github): add PR title valid format in template

</details>

### CODE REVIEW

File: jobs/etl_jobs/external/dms/main.py [hunk 1]
```diff
@@ -1,55 +1,62 @@
-import json
 import time
```
Comment (dcuesta-pass) [inline jobs/etl_jobs/external/dms/main.py:1]: /gpt

File: jobs/etl_jobs/external/dms/constants.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+from utils import access_secret_data
+
+# Environment Variables
+GCP_PROJECT_ID = os.getenv("GCP_PROJECT_ID", "passculture-data-dev")
+ENV_SHORT_NAME = os.getenv("ENV_SHORT_NAME", "dev")
+DMS_TOKEN = access_secret_data(GCP_PROJECT_ID, "token_dms")
+API_URL = "https://www.demarches-simplifiees.fr/api/v2/graphql"
+
+# Procedure IDs
+demarches = {
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/constants.py:12]: * Vu que c'est une constante => CONSTANT_CASE
* Le code étant en anglais => `PROCEDURES` ?

File: jobs/etl_jobs/external/dms/constants.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    {"name": "application_status", "type": "STRING"},
+    {"name": "last_update_at", "type": "TIMESTAMP"},
+    {"name": "application_submitted_at", "type": "TIMESTAMP"},
+    {"name": "passed_in_instruction_at", "type": "TIMESTAMP"},
+    {"name": "processed_at", "type": "TIMESTAMP"},
+    {"name": "instructors", "type": "STRING"},
+]
+
+# Additional Schema Columns for 'jeunes' and 'pro'
+fields = [
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/constants.py:33]: fields trop générique => `PRO_FIELD_NAMES`?

File: jobs/etl_jobs/external/dms/dms_query_wo_champs.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
+DMS_QUERY = """
+query getDemarches($demarcheNumber: Int!, $after: String,  $updatedSince: ISO8601DateTime) {
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/dms_query_wo_champs.py:2]: ça serait top de mettre cette query dans un fichier .graphql, et ensuite de le lire. Comme ça on a l'IDE bien adapté à la lecture de la requête

File: jobs/etl_jobs/external/dms/dms_query_w_champs.py [hunk 5]
```diff
@@ -2,7 +2,7 @@
 query getDemarches($demarcheNumber: Int!, $after: String,  $updatedSince: ISO8601DateTime) {
     demarche(number:$demarcheNumber) {
         title
-        dossiers(first: 5, after: $after, updatedSince: $updatedSince) {
+        dossiers(first:5, after: $after, updatedSince: $updatedSince) {
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/dms_query_w_champs.py:5]: idem que ci-dessous, un fichier  `.graphql` airaot été tp

File: jobs/etl_jobs/external/dms/main.py [hunk 6]
```diff
@@ -1,55 +1,58 @@
...
     print("updated_since", updated_since)
 
     if target == "jeunes":
-        fetch_dms_jeunes(updated_since, env_short_name, gcp_project_id)
+        fetch_dms(updated_since, demarches=demarches[target], target=target)
         return updated_since
 
     if target == "pro":
-        fetch_dms_pro(updated_since, env_short_name, gcp_project_id)
+        fetch_dms(updated_since, demarches=demarches[target], target=target)
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/main.py:26]: le if ne sert plus à rien

File: jobs/etl_jobs/external/dms/main.py [hunk 7]
```diff
@@ -1,55 +1,58 @@
...
-    result = fetch_result(demarches_pro, updated_since, env_short_name, gcp_project_id)
-    save_json(
-        result,
-        f"gs://data-bucket-{env_short_name}/dms_export/unsorted_dms_pro_{updated_since}.json",
-        gcp_project_id,
-    )
-
-
-def fetch_result(demarches_ids, updated_since, env_short_name, gcp_project_id):
+def fetch_result(demarches_ids, updated_since):
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/main.py:39]: :heart:

File: jobs/etl_jobs/external/dms/main.py [hunk 8]
```diff
@@ -1,55 +1,58 @@
...
     result = {}
     for demarche_id in demarches_ids:
+        print(f"Fetching demarche {demarche_id}")
+        if demarche_id in demarches["reduced"]:
+            dms_query = DMS_QUERY_REDUCED
+            print("dms query: reduced")
+        else:
+            dms_query = DMS_QUERY
+            print("dms query : default")
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/main.py:48]: top les logs! 
par contre logger.info plutot?

File: jobs/etl_jobs/external/dms/parse_dms_subscriptions_to_tabular.py [hunk 9]
```diff
@@ -5,219 +5,156 @@
...
+        or "",
+    }
+
+
+def enrich_jeunes_data(dossier, dossier_line):
+    """Enrich data for 'jeunes' target."""
+    for champ in dossier.get("champs", []):
+        if champ.get("id") == "Q2hhbXAtNTk2NDUz":
+            dossier_line["applicant_department"] = champ["stringValue"]
+        elif champ.get("id") == "Q2hhbXAtNTgyMjIx":
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/parse_dms_subscriptions_to_tabular.py:92]: si possible mettre ces 2 valeurs dans des constantes

ID_APPLICANT_DEPARTMENT
ID_APPLICANT_POSTAL_CODE

File: jobs/etl_jobs/external/dms/parse_dms_subscriptions_to_tabular.py [hunk 10]
```diff
@@ -5,219 +5,156 @@
...
-            dms_target="pro",
-            updated_since=updated_since,
-            data_gcs_bucket_name=data_gcs_bucket_name,
-        )
-    return
+def load_json_from_gcs(bucket_name, file_path):
+    """Load JSON file from Google Cloud Storage."""
+    fs = gcsfs.GCSFileSystem(project=GCP_PROJECT_ID)
+    with fs.open(f"gs://{bucket_name}/{file_path}") as json_file:
+        return json.load(json_file)
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/parse_dms_subscriptions_to_tabular.py:25]: peut être  à mettre dans dms/utils.py (avec le save_json)

File: jobs/etl_jobs/external/dms/utils.py [hunk 11]
```diff
@@ -61,3 +13,23 @@ def access_secret_data(project_id, secret_id, version_id="latest", default=None)
         return response.payload.data.decode("UTF-8")
     except DefaultCredentialsError:
         return default
+
+
+def mergeDictionary(dict_1, dict_2):
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/utils.py:18]: Ajoute une doc: `Warning: function not symetrical: dict_2 cannot contain any list)

File: jobs/etl_jobs/external/dms/utils.py [hunk 12]
```diff
@@ -61,3 +13,23 @@ def access_secret_data(project_id, secret_id, version_id="latest", default=None)
...
+    dict_3 = {**dict_1, **dict_2}
+    for key, value in dict_3.items():
+        if key in dict_1 and key in dict_2:
+            if isinstance(dict_1[key], list):
+                list_value = [value]
+                dict_3[key] = list_value + dict_1[key]
```
Comment (lmontier-pass) [inline jobs/etl_jobs/external/dms/utils.py:24]: if value is list => raise

**Review by lmontier-pass — APPROVED** [review]

**Pull request merged**